src/plugins/profiling/server/search/strategy.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { SearchTotalHits } from '@elastic/elasticsearch/lib/api/types';
+import { IEsSearchRequest, ISearchStrategy, PluginStart } from '../../../data/server';
+import {
+  autoHistogramSumCountOnGroupByField,
+  newProjectTimeQuery,
+  ProjectTimeQuery,
+} from '../routes/mappings';
+import { downsampledIndex, getSampledTraceEventsIndex } from '../routes/search_flamechart';
+import { DownsampledRequest, DownsampledTopNResponse } from '../../common/types';
+
+export const DownsampledTopNFactory = (
+  data: PluginStart
+): ISearchStrategy<DownsampledRequest, DownsampledTopNResponse> => {
+  const es = data.search.getSearchStrategy();
+
+  // FIXME these 2 constants should be configurable?
+  const initialExp = 6;
+  const targetSampleSize = 20000; // minimum number of samples to get statistically sound results
+
+  // Calculate the right down-sampled index to query data from
+  const sampleCountFromInitialExp = (filter: ProjectTimeQuery, options, deps): number => {
+    // By default, we return no samples and use the un-sampled index
+    let sampleCount = 0;
+    es.search(
+      {
+        params: {
+          index: downsampledIndex + initialExp,
```

Comment:
`downsampleIndex` sounds as if it holds an index name, which is not true - it only holds the prefix. To help understanding the code, I suggest using a function `getDownsampledIndex(initialExp)`.